### PR TITLE
osd: EC Overwrites

### PR DESCRIPTION
**Pull request by athanatos**
It's passing enough tests now to be worth looking at. For the up-to-the-minute version, see https://github.com/athanatos/ceph/tree/wip-ec-cache.

You'll probably want to start with the updated docs to get an overview. Then, the series has two pieces:
1) refactor everything to use the new and improved PGTransaction
2) use the new structure to make overwrites work

### CODE REVIEW

**Comment by athanatos** [comment]
qa-suite: https://github.com/ceph/ceph-qa-suite/pull/1227

**Comment by athanatos** [comment]
restest this please

**Comment by athanatos** [comment]
http://pulpito.ceph.com/samuelj-2016-11-01_20:17:04-rados-wip-sam-working---basic-smithi/

None of the failures seem to be relevant.

**Comment by athanatos** [comment]
Whoops, I seem to have left a bug in the snap trimmer, fixing.

File: src/osd/PGTransaction.h [hunk 1]
```diff
@@ -0,0 +1,540 @@
...
+
+    bool is_delete() const {
+      return boost::get<Init::None>(&init_type) != nullptr && delete_first;
+    }
+    bool is_none() const {
+      return boost::get<Init::None>(&init_type) != nullptr && !delete_first;
+    }
+    bool is_fresh_object() const {
+      return boost::get<Init::None>(&init_type) == nullptr;
+    }
```
Comment (liewegas) [inline src/osd/PGTransaction.h]: This whole section has me lost.  I can't figure out how None, Create, Clone, and Rename are mutually exclusive.  Unless None means the object (might) already exist.  In which case is_none() vs is_delete() is confusing.

Comment (athanatos) [inline src/osd/PGTransaction.h]: is_none() && is_delete() indicates that we are deleting an object which already exists and not recreating it.  delete_first means that the transaction logically removes the object.

There are really 4 cases:

1) We are modifying an existing object (is_none() && !is_delete()) -- rollforward
2) We are removing and then recreating an object (!is_none() && is_delete()) -- stash, rollback
3) We are removing an object (is_none() && is_delete()) -- stash, rollback
4) We are creating an object (!is_none() && !is_delete()) -- create (no stash), rollback

Create, Clone, Rename are the three ways we can recreate it.  It's confusing, but the ECBackend transaction planning needs this context to figure out how to perform the transaction.

File: src/osd/PGLog.h [hunk 2]
```diff
@@ -1104,6 +1134,9 @@ struct PGLog : DoutPrefixProvider {
     log.head = info.last_update;
     log.reset_rollback_info_trimmed_to_riter();
 
+    // TODO SAM DNM: rollforward log entries will be problematic since the on-disk
```
Comment (liewegas) [inline src/osd/PGLog.h]: dnm

File: src/osd/PG.cc [hunk 3]
```diff
@@ -1847,6 +1847,16 @@ void PG::activate(ObjectStore::Transaction& t,
 
     state_set(PG_STATE_ACTIVATING);
   }
+  if (acting.size() >= pool.info.min_size) {
+    /* TODOSAM DNM: verify that this will apply before reads are accepted
```
Comment (liewegas) [inline src/osd/PG.cc]: dnm

File: src/osd/ECTransaction.h [hunk 4]
```diff
@@ -20,25 +20,201 @@
...
+	ECUtil::HashInfoRef hinfo = get_hinfo(i.first);
+	plan.hash_infos[i.first] = hinfo;
+
+	uint64_t projected_size =
+	  hinfo->get_projected_total_logical_size(sinfo);
+
+	if (requires_rollforward(projected_size, i.second)) {
+	  plan.rollforward = true;
+	} else if (requires_inplace(projected_size, i.second)) {
+	  assert(!requires_rollforward(projected_size, i.second));
```
Comment (liewegas) [inline src/osd/ECTransaction.h]: redundant given if/else

File: src/osd/PGLog.h [hunk 5]
```diff
@@ -420,7 +420,11 @@ struct PGLog : DoutPrefixProvider {
     }
 
     // actors
-    void add(const pg_log_entry_t& e) {
+    void add(const pg_log_entry_t& e, bool applied = true) {
+      if (!applied) {
+	get_can_rollback_to() == head;
```
Comment (liewegas) [inline src/osd/PGLog.h]: s/==/=/?

**Comment by athanatos** [comment]
Removed the DNMs and fixed a bug I introduced yesterday.

File: src/osd/ReplicatedPG.cc [hunk 6]
```diff
@@ -3773,8 +3775,29 @@ ReplicatedPG::OpContextUPtr ReplicatedPG::trim_object(const hobject_t &coid)
   return ctx;
 }
 
+void ReplicatedPG::kick_snap_trim()
+{
+  assert(is_active());
+  assert(is_primary());
+  if (is_clean() && snap_trimq.empty()) {
```
Comment (liewegas) [inline src/osd/ReplicatedPG.cc]: !snap_trimq.empty()

Comment (athanatos) [inline src/osd/ReplicatedPG.cc]: Fixed

File: src/osd/ExtentCache.h [hunk 7]
```diff
@@ -0,0 +1,651 @@
...
+   2) Per-operation metadata is fixed size.
+
+   This is simple enough to realize with two main structures:
+   - extent: contains a pointer to the pin owning it and intrusive list
+             pointers to other extents owned by the same pin
+   - pin_state: contains the list head for extents owned by it
+
+   This works as long as we only need to remember one "owner" for
+   each extent.  To make this work, we'll need to leverage some
+   invariants guarranteed by higher layers:
```
Comment (manali26) [inline src/osd/ExtentCache.h]: guarranteed -> guaranteed

File: src/osd/ExtentCache.h [hunk 8]
```diff
@@ -0,0 +1,651 @@
...
+   participating in one of two pipelines at any particular point:
+
+   Write: [WaitRead -> ] WaitCommit -> WaitApply -> Complete
+   Read: WaitRead -> Complete
+
+   [WaitRead -> ] for Write indicates that we may need to start by
+   reading the full extents for any partial modifications.
+   Invariant 1) above actually indicates that we can't have writes
+   bypassing the WaitRead state while there are writes waiting on
+   Reads.  Thus, the set of operations pinning a particular extent
```
Comment (manali26) [inline src/osd/ExtentCache.h]: You might want to remove extra space before 'Thus'

File: src/osd/ECBackend.cc [hunk 9]
```diff
@@ -527,11 +589,12 @@ void ECBackend::continue_recovery_op(
 	this,
 	op.hoid,
 	op.recovery_progress.data_recovered_to,
-	recovery_max_chunk,
+	amount,
 	to_read,
-	op.recovery_progress.first);
-      op.extent_requested = make_pair(op.recovery_progress.data_recovered_to,
-				      recovery_max_chunk);
+	op.recovery_progress.first && !op.obc);
```
Comment (manali26) [inline src/osd/ECBackend.cc]: alignment

File: src/osd/ECBackend.cc [hunk 10]
```diff
@@ -40,14 +40,10 @@ struct ECRecoveryHandle : public PGBackend::RecoveryHandle {
...
-  case ECBackend::pipeline_state_t::INPLACE:
-    return lhs << "INPLACE";
-  case ECBackend::pipeline_state_t::ROLLFORWARD:
-    return lhs << "ROLLFORWARD";
-  case ECBackend::pipeline_state_t::EXCL:
-    return lhs << "EXCL";
+  case ECBackend::pipeline_state_t::CACHE_VALID:
+    return lhs << "CACHE_VALID";
+  case ECBackend::pipeline_state_t::CACHE_INVALID:
+    return lhs << "CACH_INVALID";
```
Comment (liewegas) [inline src/osd/ECBackend.cc]: typo

**Comment by athanatos** [comment]
Yeah, I'm going to tidy up the commits once the test passes on the rebased version and you've had a chance to look it over.

**Comment by liewegas** [comment]
K, I've looked it over.  I think I'll have a better time looking at the rebased version

File: src/osd/ECBackend.cc (hunk 9 above)
Comment (athanatos) [inline src/osd/ECBackend.cc]: Looks right to me.

File: src/osd/ExtentCache.h [hunk 11]
```diff
@@ -0,0 +1,651 @@
...
+      - Some read with reqid <= N is currently fetching data for
+        other extents, and this extent was in state Pending when
+	it started.
+      - All ops pinning this extent are reads in the WaitRead state
+        of the Read pipeline.
+      - There can be no unapplied writes in this case (as above).
+      - The extent must have transitioned here from Pending.
+
+
+   States 3 and 4 require a bit of explanation.  Writes and reads have
```
Comment (athanatos) [inline src/osd/ExtentCache.h]: It's aligned consistently.

File: src/osd/ExtentCache.h (hunk 8 above)
Comment (athanatos) [inline src/osd/ExtentCache.h]: Two spaces after period is a valid choice, and it's used consistently in this file.

**Comment by athanatos** [comment]
@liewegas Updated

**Comment by athanatos** [comment]
Found some bugs with PGLog, fixing.

**Comment by athanatos** [comment]
http://pulpito.ceph.com/samuelj-2016-11-15_16:33:23-rados-wip-sam-working---basic-smithi/

Getting closer, I think I just added a fix for the pg log warning.  The two hung runs seem to be btrfs hangs.  The other 3 seem to be known.  Rerunning.

**Comment by athanatos** [comment]
http://pulpito.ceph.com/samuelj-2016-11-16_21:33:31-rados-wip-sam-working---basic-smithi/

I think the rados bench one is once again a btrfs hang (seems to be reproducible).
The store_test failure doesn't seem to be related (started a master run to check).

I think we're good to merge.

@liewegas

File: src/osd/PGTransaction.h [hunk 12]
```diff
@@ -0,0 +1,561 @@
...
+     * is_none() && is_delete() indicates that we are deleting an
+     * object which already exists and not recreating it. delete_first means
+     * that the transaction logically removes the object.
+
+     * There are really 4 cases:
+
+     * 1) We are modifying an existing object (is_none() &&
+     *    !is_delete()) -- rollforward (actually, with the current
+     *    implementation, it'll still use the other rollback machinery as long
+     *    as there aren't overwrites).
```
Comment (liewegas) [inline src/osd/PGTransaction.h]: update comment here?

File: src/messages/MOSDRepOp.h [hunk 13]
```diff
@@ -55,7 +55,7 @@ class MOSDRepOp : public Message {
 
   // piggybacked osd/og state
   eversion_t pg_trim_to;   // primary->replica: trim to here
-  eversion_t pg_trim_rollback_to;   // primary->replica: trim rollback
+  eversion_t pg_roll_forward_to;   // primary->replica: trim rollback
```
Comment (liewegas) [inline src/messages/MOSDRepOp.h]: does this rename still make sense?

Comment (athanatos) [inline src/messages/MOSDRepOp.h]: Yeah, a different patch updated the terminology in PGLog to distinguish between discarding the rollback information and rolling forward.  Operationally there's no difference, but it makes explaining what's going on easier.

File: src/osd/ECBackend.cc [hunk 14]
```diff
@@ -510,8 +530,21 @@ void ECBackend::continue_recovery_op(
...
       assert(!op.recovery_progress.data_complete);
       set<int> want(op.missing_on_shards.begin(), op.missing_on_shards.end());
+      uint64_t from = op.recovery_progress.data_recovered_to;
+      uint64_t amount = get_recovery_chunk_size();
+
+      if (op.recovery_progress.first && op.obc) {
+	/* It's not in general safe to request attrs if we have the obc
+	 * cached since there might be a pending roll-forward which is
+	 * not yet reflected on the head object.  Besides, we've got the
+	 * attrs and the hinfo, might as well use them */
```
Comment (liewegas) [inline src/osd/ECBackend.cc]: more rollforward

**Pull request merged**